### PR TITLE
Show list of variants with links instead of awarded points history

### PR DESCRIPTION
**Pull request by ffund**
Currently, if a student submits a (not single variant) HW question, and has no more tries for that variant, once they click away they cannot see that submission. When they open the question again, they see a new variant. However, students like to see their own previous submission, either to help with exam review, or to revisit incorrect attempts.

This PR will show students a list of clickable previous variants in place of the current "awarded points" history list.

(Will resolve https://github.com/PrairieLearn/PrairieLearn/issues/6031, resolve https://github.com/PrairieLearn/PrairieLearn/issues/5763, resolve https://github.com/PrairieLearn/PrairieLearn/issues/3158, and resolve https://github.com/PrairieLearn/PrairieLearn/issues/29)

### CODE REVIEW

**Comment by ffund** [comment]
To do:

[ ] Show only variants that have been graded, instead of all variants
[ ] Show point value instead of attempt number
[ ] If there have been an unreasonable number of attempts, only show some + click to reveal the rest to avoid UI madness

**Comment by jonatanschroeder** [comment]
@ffund just FYI, in previous discussion I mentioned about doing this in #5909 , but this PR is now pretty huge with other aspects, so it probably makes sense to make it separately. So if you still want to proceed with this PR go for it. Sorry for the confusion.

**Comment by ffund** [comment]
> @ffund just FYI, in previous discussion I mentioned about doing this in #5909 , but this PR is now pretty huge with other aspects, so it probably makes sense to make it separately. So if you still want to proceed with this PR go for it. Sorry for the confusion.

Great, thanks for the update!

If you have any suggestion on the discussion post https://github.com/PrairieLearn/PrairieLearn/discussions/6106  I'd love to hear them :)

**Comment by ffund** [comment]
Show badges (with non-judgmental gray color) that link to previous variants on homework assessments.

In the assessment overview:

prev-variants-hw-page

On the question page:

prev-variants-question-page

**Review by nwalters512 — APPROVED** [review]
I tested this out, it's looking good to me! @mwest1066 did you want to take a final look before we land this?

File: pages/partials/questionAwardedPoints.ejs
Comment (mwest1066) [inline pages/partials/questionAwardedPoints.ejs:7]: As we discussed, I'm a bit worried about what happens when a student builds up a lot of previous variants. Here's an attempt to fix this by only showing the most recent 10 variants and providing a "..." clickable link that will reveal the earlier ones. Thoughts?

image

New `questionAwardedPoints.ejs`:
```
<% include('pointsFormatter'); %>

<% const MAX_DISPLAYED_VARIANTS = 10; %>

<% if (previous_variants) { %>
    <%
        const collapseId = `variants-points-collapse-${instance_question_id}`;
        const collapseButtonId = `variants-points-collapse-button-${instance_question_id}`;
        let startHiddenIndex = null;
        let endHiddenIndex = null;
    %>
    <% if (previous_variants.length > MAX_DISPLAYED_VARIANTS) { %>
        <a id="<%= collapseButtonId %>" role="button" onclick="$(`#<%= collapseId %>`).css('display','inline'); $(`#<%= collapseButtonId %>`).css('display','none');">&hellip;</a>
        <% startHiddenIndex = 0; %>
        <% endHiddenIndex = previous_variants.length - MAX_DISPLAYED_VARIANTS; %>
    <% } %>
    <% for (var i = 0; i < previous_variants.length; i++) { %>
        <% if (i == startHiddenIndex) { %>
            <span id="<%= collapseId %>" style="display:none">
        <% } %>
        <% if (i == endHiddenIndex) { %>
            </span>
        <% } %>
        <span class="badge badge-secondary">
            <a class="text-white" href="<%= urlPrefix %>/instance_question/<%= instance_question_id %>/?variant_id=<%= previous_variants[i].variant_id %>"><%= Math.round(previous_variants[i].max_submission_score * 100) + "%"  %></a> 
        </span>
    <% } %>
<% } %>
```

File: pages/partials/questionScorePanel.ejs
Comment (mwest1066) [inline pages/partials/questionScorePanel.ejs:14]: This ends up being a really big vertical list. I'm wondering whether we should render this as a single cell using `colspan="2"` to let it use the full table width for wrapping?

Current style:
image

Proposed change:
```
      <% if (assessment.type == 'Homework') { %>
      <tr>
        <td colspan="2" style="white-space: normal;">
          Previous variants:
          <%- include('../partials/questionAwardedPoints', {awarded_points: instance_question.variants_points_list, init_points: assessment_question.init_points, instance_question_id: instance_question.id, previous_variants: instance_question_info.previous_variants});  %>
        </td>
      </tr>
      <% } %>
```

With the proposed change:
image

File: pages/studentAssessmentInstanceHomework/studentAssessmentInstanceHomework.ejs
Comment (mwest1066) [inline pages/studentAssessmentInstanceHomework/studentAssessmentInstanceHomework.ejs:74]: Why don't we need the `colspan` and `rowspan` attributes anymore? Or said another way, why did we used to need them?

**Comment by SethPoulsen** [comment]
What are we waiting on for this to be merged? I just had a student ask for it so I came to find this PR.

**Comment by SethPoulsen** [comment]
Do we just need to implement what @mwest1066 suggested in his latest comment then it will be good to go?

**Comment by mwest1066** [comment]
Yep, if someone could just implement my proposed changes then we can merge this. I meant to do it but it keeps failing to get to the top of my ToDo list 😭

**Comment by jonatanschroeder** [comment]
> Yep, if someone could just implement my proposed changes then we can merge this. 

Not sure which is more work: implementing the changes or handling the merge conflicts that accumulated over time...

File: pages/partials/questionAwardedPoints.ejs
Comment (mwest1066) [reply pages/partials/questionAwardedPoints.ejs]: Done in 94d9da247170bc5fee7b100a2a0cd39b3d11bcee

File: pages/partials/questionScorePanel.ejs
Comment (mwest1066) [reply pages/partials/questionScorePanel.ejs]: Done in 94d9da247170bc5fee7b100a2a0cd39b3d11bcee

File: pages/studentAssessmentInstanceHomework/studentAssessmentInstanceHomework.ejs
Comment (mwest1066) [reply pages/studentAssessmentInstanceHomework/studentAssessmentInstanceHomework.ejs]: This no longer seems to be relevant in the current diff.

**Review by mwest1066 — APPROVED** [review]

File: apps/prairielearn/src/pages/partials/questionAwardedPoints.ejs
Comment (jonatanschroeder) [inline apps/prairielearn/src/pages/partials/questionAwardedPoints.ejs:26]: I believe the `a` should be outside the `span`, otherwise only the text of the badge is a link, not the badge itself.

File: apps/prairielearn/src/pages/partials/questionScorePanel.ejs
Comment (jonatanschroeder) [inline apps/prairielearn/src/pages/partials/questionScorePanel.ejs:17]: Is the `white-space: normal` necessary here?

File: apps/prairielearn/src/pages/partials/questionAwardedPoints.ejs
Comment (jonatanschroeder) [inline apps/prairielearn/src/pages/partials/questionAwardedPoints.ejs:13]: Nit: could this logic be moved to a separate script tag for cleanness? And in that case, either switch to vanilla JS, or use JQuery show/hide.

File: apps/prairielearn/src/pages/studentAssessmentInstance/studentAssessmentInstance.sql
Comment (jonatanschroeder) [inline apps/prairielearn/src/pages/studentAssessmentInstance/studentAssessmentInstance.sql:52]: Similar comments as before about aq.deleted_at.

```suggestion
WITH
  variant_max_submission_scores AS (
    SELECT
      v.id AS variant_id,
      max(s.score) AS max_submission_score
    FROM
      instance_questions AS iq
      JOIN variants AS v ON (v.instance_question_id = iq.id)
      JOIN submissions AS s ON (s.variant_id = v.id)
    WHERE
      iq.assessment_instance_id = $assessment_instance_id
      AND s.score IS NOT NULL
    GROUP BY
      v.id
  ),
  instance_question_variants AS (
    SELECT
      iq.id AS instance_question_id,
      jsonb_agg(
        jsonb_build_object(
          'variant_id',
          v.id,
          'max_submission_score',
          COALESCE(vmss.max_submission_score, 0)
        )
        ORDER BY
          v.date
      ) AS variants
    FROM
      instance_questions AS iq
      JOIN variants AS v ON (v.instance_question_id = iq.id)
      LEFT JOIN variant_max_submission_scores AS vmss ON (vmss.variant_id = v.id)
    WHERE
      iq.assessment_instance_id = $assessment_instance_id
      AND NOT v.open
      AND NOT v.broken
    GROUP BY
      iq.id
  )
```

File: apps/prairielearn/src/pages/partials/questionAwardedPoints.ejs
Comment (jonatanschroeder) [reply apps/prairielearn/src/pages/partials/questionAwardedPoints.ejs:13]: Also, use `button class="btn"` instead of `a role="button"`.

File: apps/prairielearn/src/pages/partials/questionAwardedPoints.ejs
Comment (mwest1066) [reply apps/prairielearn/src/pages/partials/questionAwardedPoints.ejs]: Good point, it should just be a single `<a class="badge">`. Fixed in f4793b734a97d9fe1002466a3c3e770101f697d3

File: apps/prairielearn/src/middlewares/selectAndAuthzInstanceQuestion.sql
Comment (mwest1066) [reply apps/prairielearn/src/middlewares/selectAndAuthzInstanceQuestion.sql]: Great point, apparently I never read these queries in my review 🤦. Applied in 029783f37

File: apps/prairielearn/src/middlewares/selectAndAuthzInstanceQuestion.sql
Comment (mwest1066) [reply apps/prairielearn/src/middlewares/selectAndAuthzInstanceQuestion.sql]: Applied in 108c3b4a1

File: apps/prairielearn/src/pages/studentAssessmentInstance/studentAssessmentInstance.sql
Comment (mwest1066) [reply apps/prairielearn/src/pages/studentAssessmentInstance/studentAssessmentInstance.sql]: Applied in ca475996c

File: apps/prairielearn/src/pages/partials/questionScorePanel.ejs
Comment (mwest1066) [reply apps/prairielearn/src/pages/partials/questionScorePanel.ejs:17]: Yep, without it you end up with this:

image

File: apps/prairielearn/src/pages/partials/questionAwardedPoints.ejs
Comment (mwest1066) [reply apps/prairielearn/src/pages/partials/questionAwardedPoints.ejs]: I replaced the jQuery with plain JS in 964127603c9aec23044d8e583b6105d91383b545. I left the code in the `onclick` handler because it's quite short and self-contained. In the future we can move this to an asset when we convert this all to `html` templates.

Regarding the use of a button, I actually deliberately chose to style this as a link instead of a button because a button takes up too much vertical space, expanding the row height (and a `btn-xs` uses a too-small font).

File: apps/prairielearn/src/pages/partials/questionScorePanel.ejs
Comment (jonatanschroeder) [reply apps/prairielearn/src/pages/partials/questionScorePanel.ejs:17]: What about adding [the text-wrap class](https://getbootstrap.com/docs/4.5/utilities/text/#text-wrapping-and-overflow)?

File: apps/prairielearn/src/pages/partials/questionAwardedPoints.ejs
Comment (jonatanschroeder) [reply apps/prairielearn/src/pages/partials/questionAwardedPoints.ejs]: What about btn-link? https://getbootstrap.com/docs/4.6/components/buttons/#examples

Comment (mwest1066) [reply apps/prairielearn/src/pages/partials/questionAwardedPoints.ejs]: That still includes all the usual button padding and margin. I think this is a case where using an actual `<a>` link but indicating that it should function as a button (`role="button"`) is cleanest?

File: apps/prairielearn/src/pages/partials/questionScorePanel.ejs
Comment (mwest1066) [reply apps/prairielearn/src/pages/partials/questionScorePanel.ejs]: Good call, that does exactly the same thing. Changed to `.text-wrap` in 4009693c923a4ca707d4aff75f3f5a9dedd392ef

File: apps/prairielearn/src/pages/partials/questionAwardedPoints.ejs
Comment (mwest1066) [reply apps/prairielearn/src/pages/partials/questionAwardedPoints.ejs]: Even using `btn btn-link p-0 m-0 border-0 rounded-0` I still couldn't get the button to preserve the line height. I think the button element itself is slightly too high (and using `btn-xs` shrinks the font). I thus left it as a `<a role="button">`.

Comment (jonatanschroeder) [reply apps/prairielearn/src/pages/partials/questionAwardedPoints.ejs]: The ellipsis don't have any visual indication that they can be clicked with the a styling, that looks weird to me. I tried with `badge-light`, it looks a little bit better, but I'll leave that decision to you. And I also wasn't able to get the button to style in a decent way.

One thing I tried that made it look a bit better was to replace `&ellip;` with `ctdot;`. Up to you.

File: apps/prairielearn/src/pages/partials/questionAwardedPoints.ejs
Comment (jonatanschroeder) [inline apps/prairielearn/src/pages/partials/questionAwardedPoints.ejs:16]: Nit: the usual suggestion I see in this case to set display to `''` (blank), so it takes the default display of the context.

**Review by jonatanschroeder — APPROVED** [review]
As I was testing this, I noticed an interesting thing that I'm not sure if it should be addressed or not. When I click on a previous variant, the button for "Try a new variant" would show. This button would create a new variant if no open variant exists, but if an open/incomplete variant exists, it would switch to the open variant (i.e., a variant that has no submission, or one that has been saved but not graded, or one where the submission is ungradable, or one with multiple tries per variant). I believe this behaviour is expected and should remain, but the text of the button ("Try a new variant") is misleading. Do you think it's worth the effort to change it? In other words, if there is an open variant in the current instance question that is not the current one, change the text to something that says "Go to latest variant" or equivalent?

Also, that latest variant is never shown in the history. Should it? At least if it has a submission? In that case it could show "Open", or if it has a valid submission (e.g., >1 tries per submission) the percentage of the submissions so far plus "/Open"? Possibly with a different colour?

I'm approving the PR in the case where you believe the comments above should not be addressed in the current PR, but if you decide to address them here feel free to request a new review.

File: apps/prairielearn/src/pages/partials/questionAwardedPoints.ejs
Comment (mwest1066) [reply apps/prairielearn/src/pages/partials/questionAwardedPoints.ejs]: I switched to a `<button>` by avoiding Bootstrap's `.btn` styling and explicitly styling a built-in button. This gives the correct spacing.

I agree with you that I'd prefer a better visual indication that this element is clickable. It does have the mouse pointer change on hover, which helps for platforms with pointers but not for touch platforms. I tried using a `.badge` but that doesn't play well with `<button>` and I'd rather not fall back to `<a>` now that I've got `<button>` working (for better accessibility support).

I like the idea of `&ctdot;` -- I switched to that.

File: apps/prairielearn/src/pages/partials/questionAwardedPoints.ejs
Comment (mwest1066) [reply apps/prairielearn/src/pages/partials/questionAwardedPoints.ejs]: Oh nice! TIL! Changed in 78fe96881285365ac201672163f65482e41d1ddc

**Comment by mwest1066** [comment]
> As I was testing this, I noticed an interesting thing that I'm not sure if it should be addressed or not. When I click on a previous variant, the button for "Try a new variant" would show. This button would create a new variant if no open variant exists, but if an open/incomplete variant exists, it would switch to the open variant (i.e., a variant that has no submission, or one that has been saved but not graded, or one where the submission is ungradable, or one with multiple tries per variant). I believe this behaviour is expected and should remain, but the text of the button ("Try a new variant") is misleading. Do you think it's worth the effort to change it? In other words, if there is an open variant in the current instance question that is not the current one, change the text to something that says "Go to latest variant" or equivalent?
> 
> Also, that latest variant is never shown in the history. Should it? At least if it has a submission? In that case it could show "Open", or if it has a valid submission (e.g., >1 tries per submission) the percentage of the submissions so far plus "/Open"? Possibly with a different colour?
> 
> I'm approving the PR in the case where you believe the comments above should not be addressed in the current PR, but if you decide to address them here feel free to request a new review.

Thanks. These are all reasonable questions, but I'm going to merge this PR as is because I think it brings useful functionality even in its current state and I don't have the bandwidth to devote to PR anymore (I was only trying to get @ffund's original PR merged). I'd be very happy to discuss the questions above or review a followup PR.

**Pull request merged**
